test/passes/Os_print-stack-ir.txt

```diff
@@ -50,15 +50,23 @@
  (export "ppf64" (func $3))
  (func $0 (; 0 ;) (result i32)
   i32.const 1
+  push
```

Comment:
Interesting that this is the only place where we see the push+pops from a print-stack-ir pass. I assume this is because Os isn't coalescing these non-stacky push+pops into exprs, where it is everywhere else?

Essentially, this shouldn't unnecessarily triple the line noise for every i32.const when printing (optimized) stack IR. So, seems good.

Reply:
> I assume this is because Os isn't coalescing these non-stacky push+pops into exprs, where it is everywhere else?

I don't think there are pushes or pops anywhere else to begin with. The pushes and pops here are explicitly written into the .wast file.

Reply:
@tlively is right; they are [written](https://github.com/WebAssembly/binaryen/blob/3fbec348f6de9c213d0091984305f9bb4906af47/test/passes/Os_print-stack-ir.wast#L23-L24) in the source `wast` file. Actually we don't generate `push` anywhere in Binaryen now unless it is parsed from an input wast file. The only occasion we generate `pop` in Binaryen is [here](https://github.com/WebAssembly/binaryen/blob/3fbec348f6de9c213d0091984305f9bb4906af47/src/wasm/wasm-binary.cpp#L2394) and this is for `exnref.pop`, a pseudo instruction that provides `exnref` value produced by `catch` instruction.

---

Background on `catch`:
In real semantics, `catch` instruction pushes an `exnref` value onto the value stack, but if we model `catch` instruction that way, the wast file and the AST would look a bit confusing, because `catch` instruction would be buried deep down in some children AST node. For example, suppose we have this wat file:
```wat
try
  ...
catch
  local.set 0
  ...
```

If we model `catch` as it returns an `exnref`, the wast will be like
```wast
(try
  ...
  (local.set 0
    (catch)
  )
)
```
So it wouldn't be very clear that catch divides kind of a new section within the try-catch-end structure. So when we implemented `catch`, we decided to make it resemble if-else-end, and make `catch` as something like a barrier (like `else`) that does not return any value. Instead, A pseudo instruction `exnref.pop` will follow right after `catch` and it will push the `exnref` value onto the stack.

This discussion was back in Jun 2019 or something, and we had different plans to use `push` and `pop` for other usages as well then, but now I'm not sure if we will need these instructions other than this purpose.